ghidra/p_code_extractor/PcodeExtractor.java

```diff
@@ -0,0 +1,1091 @@
+
+        return call;
+    }
+
+
+    /**
+     * @param pCodeCount: Pcode index in current block
+     * @param pcodeOp:    Pcode instruction
+     * @param instrAddr:  Assembly instruction address
+     * @return: new Def Term
+     * 
+     * Creates a Def Term with an unique TID consisting of the prefix def, its instruction address and the index of the pcode in the block.
+     */
+    protected Term<Def> createDefTerm(int pcodeIndex, PcodeOp pcodeOp, Address instrAddr) {
+        Tid defTid = new Tid(String.format("instr_%s_%s", instrAddr.toString(), pcodeIndex), instrAddr.toString());
+        if (pcodeOp.getMnemonic().equals("STORE")) {
+            return new Term<Def>(defTid, new Def(createExpression(pcodeOp), pcodeIndex));
+            // cast copy instructions that have address outputs into store instructions
+        } else if (pcodeOp.getMnemonic().equals("COPY") && pcodeOp.getOutput().isAddress()) {
+            return new Term<Def>(defTid, new Def(new Expression("STORE", null, createVariable(pcodeOp.getOutput(), false), createVariable(pcodeOp.getInput(0), false)), pcodeIndex));
+        }
+        return new Term<Def>(defTid, new Def(createVariable(pcodeOp.getOutput(), false), createExpression(pcodeOp), pcodeIndex));
+    }
+
+
+    /**
+     * @param node: Varnode source for Variable
+     * @return: new Variable
+     * 
+     * Set register name based on being a register, virtual register, constant or ram address.
+     * In case it is a virtual register, prefix the name with $U.
+     * In case it is a constant, remove the const prefix from the constant.
+     */
+    protected Variable createVariable(Varnode node, Boolean isArgument) {
+        Variable var = new Variable();
+        if (node.isRegister()) {
+            String mnemonic = getRegisterMnemonic(node);
+            if(isArgument && mnemonic.equals("AL")){
+                return castToFullRegister(var, node);
+            }
+            var.setName(mnemonic);
+            var.setIsVirtual(false);
+        } else if (node.isUnique()) {
+            var.setName(renameVirtualRegister(node.getAddress().toString()));
+            var.setIsVirtual(true);
+        } else if (node.isConstant()) {
+            var.setValue(removeConstantPrefix(node.getAddress().toString()));
+            var.setIsVirtual(false);
+        } else if (node.isAddress()) {
+            var.setAddress(node.getAddress().toString());
+            var.setIsVirtual(false);
+        } else if (node.isFree()) {
+            var.setAddress(removeStackPrefix(node.getAddress().toString()));
+            var.setIsVirtual(false);
+        }
+
+        var.setSize(node.getSize());
+
+        return var;
+    }
+
+
+    /**
+     * 
+     * @param var: register variable
+     * @param node: varnode containing half register
+     * @return: full register variable
+     * 
+     * Casts half registers to full registers
+     */
+    protected Variable castToFullRegister(Variable var, Varnode node) {
```

Comment:
I think there is a function in the Ghidra API that does this independent of the register and independent of the CPU architecture.

Reply:
Done